core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,222 @@
+
+    val kubectlBin = Try {
+      alternatives.find(a => Files.isExecutable(Paths.get(a))).get
+    } getOrElse {
+      throw new FileNotFoundException(s"Couldn't locate kubectl binary (tried: ${alternatives.mkString(", ")}).")
+    }
+
+    Seq(kubectlBin)
+  }
+
+  def run(image: String, name: String, environment: Map[String, String] = Map(), labels: Map[String, String] = Map())(
+    implicit transid: TransactionId): Future[ContainerId] = {
+    val environmentArgs = environment.flatMap {
+      case (key, value) => Seq("--env", s"$key=$value")
+    }.toSeq
+
+    val labelArgs = labels.map {
+      case (key, value) => s"$key=$value"
+    } match {
+      case Seq() => Seq()
+      case pairs => Seq("-l") ++ pairs
+    }
+
+    val runArgs = Seq(
+      "run",
+      name,
+      "--image",
+      image,
+      "--generator",
+      "run-pod/v1",
+      "--restart",
+      "Always",
+      "--limits",
+      "memory=256Mi") ++ environmentArgs ++ labelArgs
+
+    runCmd(runArgs, timeouts.run)
+      .map { _ =>
+        name
+      }
+      .map(ContainerId.apply)
+  }
+
+  def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerAddress] = {
+    Future {
+      blocking {
+        val pod =
+          kubeRestClient.pods().withName(id.asString).waitUntilReady(timeouts.inspect.length, timeouts.inspect.unit)
+        ContainerAddress(pod.getStatus().getPodIP())
+      }
+    }.recoverWith {
+      case e =>
+        log.error(this, s"Failed to get IP of Pod '${id.asString}' within timeout: ${e.getClass} - ${e.getMessage}")
+        Future.failed(new Exception(s"Failed to get IP of Pod '${id.asString}'"))
+    }
+  }
+
+  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+
+  def logs(id: ContainerId, sinceTime: String)(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    Source.fromFuture(Future {
+      blocking {
+        val sinceTimePart = if (!sinceTime.isEmpty) {
+          s"&sinceTime=${sinceTime}"
+        } else ""
+        val url = new URL(
+          URLUtils.join(
+            kubeRestClient.getMasterUrl.toString,
+            "api",
+            "v1",
+            "namespaces",
+            kubeRestClient.getNamespace,
+            "pods",
+            id.asString,
+            "log?timestamps=true" ++ sinceTimePart))
+        val request = new Request.Builder().get().url(url).build
+
+        val response = kubeRestClient.getHttpClient.newCall(request).execute
+        if (!response.isSuccessful) {
+          Future.failed(
+            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
```

Comment:
This failed `Future` will be thrown away and not used anywhere. I think what you're looking for is:

```scala
Future {
  blocking {
    // your request code
    val response = kubeRestClient.getHttpClient.newCall(request).execute
    if(response.isSuccessful) {
      Future.successful(response.body.string)
    } else {
      Future.failed(
            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
    }
  }
}.flatMap(identity).map { output => ... }
```

An `if` is an expression in Scala, meaning it'll return something (namely: the value of the last line). The `Future { }` block in your case always returns `response.body.string`. If you want to fold the error case into the future chain, you'll explicitly need to return a `Success` for the good case and a `Failure` for the bad case. Since it makes chaining easier, I used the `Future.successful` and `Future.failed` variants, which are immediatly resolved Futures of the desired state. The return type would now be `Future[Future[String]]` though, since I returned a Future inside another Future. To flatten those (flatten is only available in Scala 2.12) you can use `flatMap(identity)`. `identity` is a predefined function. You can as well write `flatMap(x => x)`.

Hope that makes sense.